### PR TITLE
Add WDAC Audit mode logging as experimental feature

### PR DESCRIPTION
**Pull request by PaulHigin**
# PR Summary

This PR adds a new language mode and logging messages to PowerShell restrictions when run under Windows WDAC (Windows Defender Application Control) policy.

This addresses Issue #18628 

## PR Context

PowerShell enforces language mode and other restrictions when it detects it is running under a WDAC (or deprecated AppLocker) system polices (Windows platform only).  The enforced restrictions can cause existing script (written without the restrictions in mind) to break in unpredictable ways.  

WDAC can run in Audit mode where policy restriction instances are not enforced but logged.  These changes updates PowerShell so that when WDAC is in Audit mode, it too logs when restrictions would apply if the policy was being enforced.  This will help users to understand what restrictions will apply when their script runs in policy enforcement mode.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [x] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [x] Experimental feature name(s): PSWDACAuditLogging
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 1]
```diff
@@ -3977,7 +3978,9 @@ internal ConvertViaNoArgumentConstructor(ConstructorInfo constructor, Type type)
                     //  - It's in FullLanguage but not because it's part of a parameter binding that is transitioning from ConstrainedLanguage to FullLanguage
                     // When this is invoked from a parameter binding in transition from ConstrainedLanguage environment to FullLanguage command, we disallow
                     // the property conversion because it's dangerous.
-                    if (ecFromTLS == null || (ecFromTLS.LanguageMode == PSLanguageMode.FullLanguage && !ecFromTLS.LanguageModeTransitionInParameterBinding))
+                    if (ecFromTLS == null || 
+                        ((ecFromTLS.LanguageMode == PSLanguageMode.FullLanguage || ecFromTLS.LanguageMode == PSLanguageMode.ConstrainedLanguageAudit) &&
+                         !ecFromTLS.LanguageModeTransitionInParameterBinding))
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs:3983]: This scenario is: the parameter binding happens in constrained language mode for a command that is about to run in the full language mode (`LanguageModeTransitionInParameterBinding` is set to `true` in this scenario).

File: src/System.Management.Automation/engine/ParameterBinderBase.cs [hunk 2]
```diff
@@ -1244,8 +1244,9 @@ internal InternalCommand Command
                         // However, we don't allow Hashtable-to-Object conversion (PSObject and IDictionary) because
                         // those can lead to property setters that probably aren't expected. This is enforced by
                         // setting 'Context.LanguageModeTransitionInParameterBinding' to true before the conversion.
+                        var currentLanguageMode = Context.LanguageMode;
                         bool changeLanguageModeForTrustedCommand =
-                            Context.LanguageMode == PSLanguageMode.ConstrainedLanguage &&
+                            currentLanguageMode == PSLanguageMode.ConstrainedLanguage &&
                             this.Command.CommandInfo.DefiningLanguageMode == PSLanguageMode.FullLanguage;
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/ParameterBinderBase.cs:1250]: Shall we check on `currentLanguageMode == PSLanguageMode.ConstrainedLanguageAudit` too? `LanguageModeTransitionInParameterBinding` should be set to `true` if we are mimicking `ConstrainedLanguage` mode.

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]
Doesn't have to be code fenced, but we should advertise this as an Experimental Feature in case we need to change the trace messages based on user feedback

File: src/System.Management.Automation/engine/ExternalScriptInfo.cs [hunk 3]
```diff
@@ -532,7 +537,7 @@ private void ReadScriptContents()
                                     {
                                         DefiningLanguageMode = Context.LanguageMode;
                                     }
-
+
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/ExternalScriptInfo.cs:540]: Space characters can be removed on this line :)

File: src/System.Management.Automation/engine/Modules/ExportModuleMemberCommand.cs [hunk 4]
```diff
@@ -167,9 +168,20 @@ protected override void ProcessRecord()
             if (Context.EngineSessionState.Module?.LanguageMode != null &&
                 Context.LanguageMode != Context.EngineSessionState.Module.LanguageMode)
             {
-                var se = new PSSecurityException(Modules.CannotExportMembersAccrossLanguageBoundaries);
-                var er = new ErrorRecord(se, "Modules_CannotExportMembersAccrossLanguageBoundaries", ErrorCategory.SecurityError, this);
-                ThrowTerminatingError(er);
+                if (Context.LanguageMode != PSLanguageMode.ConstrainedLanguageAudit && 
+                    Context.EngineSessionState.Module.LanguageMode != PSLanguageMode.ConstrainedLanguageAudit)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/Modules/ExportModuleMemberCommand.cs:172]: (note to myself) need to take a closer look at this.

**Comment by daxian-dbw** [comment]
@PaulHigin I'm closing this one assuming it's superseded by #19641

**Pull request closed without merging**